### PR TITLE
DOCK-2352: Fixed path for external files when using custom readme

### PR DESCRIPTION
**Pull request by hyunnaye**
**Description**
This PR fixes the `baseUrl` of the readme for `ngx-markdown`. Before the fix, when using a custom readme, the base path was still set to the root of the repo which resulted in 404 errors when redirecting to external files in the readme. This PR will set the `baseUrl` to the directory of the custom readme file.

For workflow: https://qa.dockstore.org/workflows/github.com/denis-yuen/myco/myco_cleaned:develop?tab=info
Before, the input.md link would be https://github.com/denis-yuen/myco/blob/develop/inputs.md leading to a 404 error. This PR changes it to https://github.com/denis-yuen/myco/blob/develop/doc/inputs.md 

Before, images would lead to a 404 error on the console. ([Workflow](https://qa.dockstore.org/workflows/github.com/denis-yuen/myco/myco_sra:main?tab=info))
Screenshot from 2023-04-12 13-13-58

It now links to https://github.com/denis-yuen/myco/blob/split-further/doc/doc_making_resources/myco_sra_flowchart_1.png thus removing the 404 errors on console. The images still do not load as they are linked to its GitHub link instead of the raw image, we have created a [ticket](https://github.com/dockstore/dockstore/issues/5459) to implement our wrapper to load our images correctly.
Screenshot from 2023-04-12 13-19-08

**Review Instructions**
Review that the basepath is now correctly set.

**Issue**
https://github.com/dockstore/dockstore/issues/5423

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that your code compiles by running `npm run build`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] If this is the first time you're submitting a PR or even if you just need a refresher, consider reviewing our [style guide](https://github.com/dockstore/dockstore/wiki/Dockstore-Frontend-Opinionated-Style-Guide#pr-checklist)
- [x] Do not bypass Angular sanitization (bypassSecurityTrustHtml, etc.), or justify why you need to do so
- [x] If displaying markdown, use the `markdown-wrapper` component, which does extra sanitization
- [x] Do not use cookies, although this may change in the future
- [x] Run `npm audit` and ensure you are not introducing new vulnerabilities
- [x] Do due diligence on new 3rd party libraries, checking for CVEs
- [x] Don't allow user-uploaded images to be served from the Dockstore domain
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.
- [x] Check whether this PR disables tests. If it legitimately needs to disable a test, create a new ticket to re-enable it in a specific milestone.

### CODE REVIEW

File: src/app/workflow/info-tab/info-tab.component.ts [hunk 1]
```diff
@@ -132,6 +133,12 @@ export class InfoTabComponent extends EntryTab implements OnInit, OnChanges {
       this.workflowsService.getWorkflowVersionOrcidAuthors(this.workflow.id, this.selectedVersion.id).subscribe((orcidAuthors) => {
         this.authors = [...this.selectedVersion.authors, ...orcidAuthors];
       });
+      if (this.selectedVersion.descriptionSource === 'CUSTOM_README') {
+        const slashIndex = this.selectedVersion.readMePath.slice(0, -1).lastIndexOf('/'); //index of the slash before the readme filename. This allows us to get the path to its parent directory.
```
Comment (svonworl) [inline src/app/workflow/info-tab/info-tab.component.ts:137]: If I understand correctly, the `slice(0, -1)` removes the last character from the string, why do we need to do that?

File: src/app/workflow/info-tab/info-tab.component.ts [hunk 2]
```diff
@@ -132,6 +133,12 @@ export class InfoTabComponent extends EntryTab implements OnInit, OnChanges {
...
+        this.readMeBasePath = this.selectedVersion.name + this.selectedVersion.readMePath.substring(0, slashIndex);
```
[the remark this answers is not in the document]
Comment (hyunnaye) [reply src/app/workflow/info-tab/info-tab.component.ts:138]: The path always starts with a slash since relative paths get converted to absolute from this PR https://github.com/dockstore/dockstore/pull/5110

**Review by coverbeck — APPROVED** [review]

**Review by svonworl — APPROVED** [review]

File: src/app/workflow/info-tab/info-tab.component.ts [hunk 3]
```diff
@@ -132,6 +133,12 @@ export class InfoTabComponent extends EntryTab implements OnInit, OnChanges {
       this.workflowsService.getWorkflowVersionOrcidAuthors(this.workflow.id, this.selectedVersion.id).subscribe((orcidAuthors) => {
         this.authors = [...this.selectedVersion.authors, ...orcidAuthors];
       });
+      if (this.selectedVersion.descriptionSource === 'CUSTOM_README') {
```
Comment (kathy-t) [inline src/app/workflow/info-tab/info-tab.component.ts:136]: I think it would be better to move this logic into the `BaseUrlPipe` so that if the pipe is used else where in the future, we're aware that the custom readme path can affect the base url. Maybe add a `customReadMePath` parameter to the pipe - this way, we can also create some tests for the custom readme path in https://github.com/dockstore/dockstore-ui2/blob/develop/src/app/shared/entry/base-url.pipe.spec.ts

File: src/app/workflow/info-tab/info-tab.component.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kathy-t) [reply src/app/workflow/info-tab/info-tab.component.ts:138]: Should add as a comment that the readme path always starts with a slash because it's not immediately obvious

File: src/app/workflow/info-tab/info-tab.component.ts (hunk 3 above)
Comment (svonworl) [reply src/app/workflow/info-tab/info-tab.component.ts:136]: A wrinkle is that notebooks can also contain markdown, which can reference images using relative URLs.  The base URL for a notebook's markdown is probably similar to that of a custom `README` (the repo info + the directory portion of the notebook path), so if we can account for that in our pipe mods, that would be great.

File: src/app/shared/entry/base-url.pipe.spec.ts [hunk 4]
```diff
@@ -5,7 +5,9 @@ describe('Pipe: BaseUrl', () => {
     const pipe = new BaseUrlPipe();
     expect(pipe).toBeTruthy();
     expect(pipe.transform('https://github.com/foobar/hello_world', 'master')).toBe('https://github.com/foobar/hello_world/blob/master/');
-    expect(pipe.transform('https://gitlab.com/foobar/hello_world', 'develop')).toBe('https://gitlab.com/foobar/hello_world/blob/develop/');
+    expect(pipe.transform('https://gitlab.com/foobar/hello_world', 'develop', 'dir/readme.md')).toBe(
```
Comment (kathy-t) [inline src/app/shared/entry/base-url.pipe.spec.ts:8]: Based on this other [comment thread](https://github.com/dockstore/dockstore-ui2/pull/1753#discussion_r1164818203), I thought the readme path always started with a slash?

Comment (kathy-t) [inline src/app/shared/entry/base-url.pipe.spec.ts:8]: I would also add a test case for when the readme path is `null`

File: src/app/shared/entry/base-url.pipe.ts [hunk 5]
```diff
@@ -4,21 +4,31 @@ import { Pipe, PipeTransform } from '@angular/core';
   name: 'baseUrl',
 })
 export class BaseUrlPipe implements PipeTransform {
+  private markdownBasePath: string;
```
Comment (kathy-t) [inline src/app/shared/entry/base-url.pipe.ts:7]: I don't think this needs to be a private field because it's only used inside `transform`. It can be a variable inside that function instead

File: src/app/shared/entry/base-url.pipe.ts [hunk 6]
```diff
@@ -4,21 +4,31 @@ import { Pipe, PipeTransform } from '@angular/core';
...
+    if (customPath) {
+      const slashIndex = customPath.lastIndexOf('/'); //index of the slash before the readme filename. This allows us to get the path to its parent directory.
+      this.markdownBasePath = customPath.substring(0, slashIndex) + '/'; // the last slash is needed by ngx-markdown
+    } else {
+      this.markdownBasePath = '';
+    }
+
     if (providerUrl.startsWith('https://github.com/') || providerUrl.startsWith('https://gitlab.com/')) {
-      return providerUrl + '/blob/' + versionName + '/'; // the last slash is needed by ngx-markdown
+      return providerUrl + '/blob/' + versionName + '/' + this.markdownBasePath; // the last slash needed by ngx-markdown is provided by this.markdownBasePath.
```
Comment (kathy-t) [inline src/app/shared/entry/base-url.pipe.ts:29]: The comment doesn't seem quite accurate. `markdownBasePath` can be `''` which doesn't have the last slash. You could remove the last slash assignment on this line and add it to line 25 where you assign `this.markdownBasePath = ''`

File: src/app/shared/entry/base-url.pipe.ts [hunk 7]
```diff
@@ -4,21 +4,31 @@ import { Pipe, PipeTransform } from '@angular/core';
...
     } else if (providerUrl.startsWith('https://bitbucket.org/')) {
-      return providerUrl + '/src/' + versionName + '/'; // the last slash is needed by ngx-markdown
+      return providerUrl + '/src/' + versionName + '/' + this.markdownBasePath; //  the last slash needed by ngx-markdown is provided by this.rmarkdownBasePath.
```
Comment (kathy-t) [inline src/app/shared/entry/base-url.pipe.ts:31]: Similar comment about the last slash

File: src/app/workflow/info-tab/info-tab.component.ts (hunk 3 above)
Comment (hyunnaye) [reply src/app/workflow/info-tab/info-tab.component.ts]: To account for notebook markdown files, I generalized the parameter name in `BaseUrlPipe` to `customPath` instead of `readMePath`

File: src/app/shared/entry/base-url.pipe.ts [hunk 8]
```diff
@@ -4,21 +4,31 @@ import { Pipe, PipeTransform } from '@angular/core';
...
+  private markdownBasePath: string;
   /**
    * Constructs a base url for relative links in markdown README files.
    * @param providerUrl
    * @param versionName
+   * @param customPath If applicable, the relative path of markdown file if its parent directory is not the root of the repo. Ex. custom readMePath defined in .dockstore.yml
    * @returns {string} base url to be used by ngx-markdown
    */
-  transform(providerUrl: string, versionName: string): string {
+  transform(providerUrl: string, versionName: string, customPath?: string): string {
```
Comment (kathy-t) [inline src/app/shared/entry/base-url.pipe.ts:15]: `customPath` can be null, so you could also add the `null` type beside `string`

File: src/app/shared/entry/base-url.pipe.spec.ts (hunk 4 above)
Comment (hyunnaye) [reply src/app/shared/entry/base-url.pipe.spec.ts:8]: Yes, it does. Sorry looks like I didn't add it on this. Thanks for catching that

File: src/app/shared/entry/base-url.pipe.ts [hunk 9]
```diff
@@ -8,17 +8,27 @@ export class BaseUrlPipe implements PipeTransform {
...
+  transform(providerUrl: string, versionName: string, customPath?: string | null): string {
+    let markdownBasePath: string;
     if (!providerUrl || !versionName) {
       console.error('providerUrl or versionName is not truthy');
       return null;
     }
+
+    if (customPath) {
+      const slashIndex = customPath.lastIndexOf('/'); //index of the slash before the readme filename. This allows us to get the path to its parent directory.
+      markdownBasePath = customPath.substring(0, slashIndex) + '/'; // the last slash is needed by ngx-markdown
```
Comment (kathy-t) [inline src/app/shared/entry/base-url.pipe.ts:23]: Can you add a comment that `customPath` always starts with a `/`?

**Review by kathy-t — APPROVED** [review]
